# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Get the probability mass function given marginal cdf's and the
#' correlation.
#'
#' @param g_cdf The cdf of the first SNP.
#' @param h_cdf The cdf of the second SNP.
#' @param rho The correlation.
#'
#' @return A matrix. The rows index the first SNP and the columns index the second SNP.
#'     Element (i, j) is the probability of genotype i - 1 in SNP 1 and j - 1 in SNP 2.
#'
#' @author David Gerard
dist_from_marg <- function(g_cdf, h_cdf, rho) {
    .Call('_uglycat_dist_from_marg', PACKAGE = 'uglycat', g_cdf, h_cdf, rho)
}

#' Likelihood of the correlation given the count data.
#'
#' @param atanh_rho The inverse hyperbolic tangent of the correlation.
#'     This is also known as the Fisher Z-transformation of the correlation.
#' @param lX A matrix with the genotype log-likelihoods for the first SNP. Element
#'     (\code{i}, \code{j}) is the log-likelihood
#'     of the data for individual \code{i} given dosage \code{j} for SNP 1.
#' @param lY A matrix with the genotype log-likelihoods for the second SNP. Element
#'     (\code{i}, \code{j}) is the log-likelihood
#'     of the data for individual \code{i} given dosage \code{j} for SNP 2.
#' @param lg A vector with the log-probabilities of the genotypes for the first SNP. Element
#'     \code{j} is the log-probability of dosage \code{j - 1} in SNP 1.
#' @param lh A vector with the log-probabilities of the genotypes for the second SNP. Element
#'     \code{j} is the log-probability of dosage \code{j - 1} in SNP 2.
#'
#' @author David Gerard
#'
#' @seealso
#' \describe{
#'   \item{\code{\link{correst}}}{For the function that optimizes \code{corrlike_r}.}
#' }
#'
corrlike <- function(atanh_rho, lX, lY, lg, lh) {
    .Call('_uglycat_corrlike', PACKAGE = 'uglycat', atanh_rho, lX, lY, lg, lh)
}

corr_optim <- function(atanh_rho, lX, lY, lg, lh) {
    .Call('_uglycat_corr_optim', PACKAGE = 'uglycat', atanh_rho, lX, lY, lg, lh)
}

#' A C++ implementation of bivariate normal probabilities, derived from the MATLAB code of Alan Genz.
#'
#' Calculates the probability that x > dh and y > dk.
#'
#' Original Matlab code written by:
#' Alan Genz, Department of Mathematics,
#' Washington State University,
#' Pullman, Wa 99164-3113, Email: alangenz@wsu.edu,
#'
#' @param dh 1st lower integration limit
#' @param dk 2nd lower integration limit
#' @param r Correlation coefficient
#'
#'
#' @references Drezner, Z., & Wesolowsky, G. O. (1990). On the computation of the bivariate normal integral. Journal of Statistical Computation and Simulation, 35(1-2), 101-107.
#'
#' @author David Gerard
bvnu <- function(dh, dk, r) {
    .Call('_uglycat_bvnu', PACKAGE = 'uglycat', dh, dk, r)
}

#' A C++ implementation of the bivariate normal CDF, derived from the MATLAB code of Alan Genz
#'
#' Calculates the probability that x < dh and y < dk
#'
#' Original Matlab code written by: Alan Genz,
#' Department of Mathematics, Washington State University,
#' Pullman, Wa 99164-3113, Email: alangenz@wsu.edu,
#'
#' @param dh 1st upper integration limit
#' @param dk 2nd upper integration limit
#' @param r Correlation coefficient
#'
#'
#' @references Drezner, Z., & Wesolowsky, G. O. (1990). On the computation of the bivariate normal integral. Journal of Statistical Computation and Simulation, 35(1-2), 101-107.
#'
#'
#' @author David Gerard
bvnl <- function(dh, dk, r) {
    .Call('_uglycat_bvnl', PACKAGE = 'uglycat', dh, dk, r)
}

#' Distribution function of the bivariate Gaussian copula.
#'
#' @param x The value of the first element. Must be between 0 and 1.
#' @param y The value of the second element. Must be between 0 and 1.
#' @param rho The correlation. Must be between 0 and 1.
#'
#' @author David Gerard
#'
#' @seealso \code{\link{bvnl}} For the cdf of the bivariate normal.
#'
#'
pnormcop <- function(x, y, rho) {
    .Call('_uglycat_pnormcop', PACKAGE = 'uglycat', x, y, rho)
}

#' Calculate the genotypes log-likelihoods.
#'
#' @param refvec A vector of reference counts.
#' @param sizevec A vector of size counts.
#' @param ploidy The ploidy of the species.
#' @param seq The sequencing error rate.
#' @param bias The allele bias parameter.
#' @param od The overdispersion parameter.
#'
#'
#' @return A matrix of numerics. The rows index the individuals and the
#'     columns index the genotype. Element (i, j) is the log-likelihood
#'     of dosage j for individual i.
#'
#' @author David Gerard
#'
#'
gen_like <- function(refvec, sizevec, ploidy, seq, bias, od) {
    .Call('_uglycat_gen_like', PACKAGE = 'uglycat', refvec, sizevec, ploidy, seq, bias, od)
}

#' Log of cumulative sum of exponential.
#'
#' @param x A vector to log-cum-sum-exp.
#'
#' @return A NumericVector. Element i is the log of the sum of
#'     the exponential of the first i elements in \code{x}.
#'
#' @author David Gerard
log_cum_sum_exp <- function(x) {
    .Call('_uglycat_log_cum_sum_exp', PACKAGE = 'uglycat', x)
}

sum_out_prods <- function(pm1, pm2, crossmat) {
    .Call('_uglycat_sum_out_prods', PACKAGE = 'uglycat', pm1, pm2, crossmat)
}

